src/main/java/org/opensearch/geospatial/ip2geo/action/UpdateDatasourceTransportAction.java

```diff
@@ -0,0 +1,184 @@
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.support.ActionFilters;
+import org.opensearch.action.support.HandledTransportAction;
+import org.opensearch.action.support.master.AcknowledgedResponse;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.geospatial.ip2geo.common.DatasourceFacade;
+import org.opensearch.geospatial.ip2geo.common.DatasourceManifest;
+import org.opensearch.geospatial.ip2geo.common.Ip2GeoLockService;
+import org.opensearch.geospatial.ip2geo.jobscheduler.Datasource;
+import org.opensearch.geospatial.ip2geo.jobscheduler.DatasourceUpdateService;
+import org.opensearch.jobscheduler.spi.schedule.IntervalSchedule;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+/**
+ * Transport action to update datasource
+ */
+@Log4j2
+public class UpdateDatasourceTransportAction extends HandledTransportAction<UpdateDatasourceRequest, AcknowledgedResponse> {
+    private static final long LOCK_DURATION_IN_SECONDS = 300l;
+    private final Ip2GeoLockService lockService;
+    private final DatasourceFacade datasourceFacade;
+    private final DatasourceUpdateService datasourceUpdateService;
+
+    /**
+     * Constructor
+     *
+     * @param transportService the transport service
+     * @param actionFilters the action filters
+     * @param lockService the lock service
+     * @param datasourceFacade the datasource facade
+     * @param datasourceUpdateService the datasource update service
+     */
+    @Inject
+    public UpdateDatasourceTransportAction(
+        final TransportService transportService,
+        final ActionFilters actionFilters,
+        final Ip2GeoLockService lockService,
+        final DatasourceFacade datasourceFacade,
+        final DatasourceUpdateService datasourceUpdateService
+    ) {
+        super(UpdateDatasourceAction.NAME, transportService, actionFilters, UpdateDatasourceRequest::new);
+        this.lockService = lockService;
+        this.datasourceUpdateService = datasourceUpdateService;
+        this.datasourceFacade = datasourceFacade;
+    }
+
+    /**
+     * Get a lock and update datasource
+     *
+     * @param task the task
+     * @param request the request
+     * @param listener the listener
+     */
+    @Override
+    protected void doExecute(final Task task, final UpdateDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        lockService.acquireLock(request.getName(), LOCK_DURATION_IN_SECONDS, ActionListener.wrap(lock -> {
+            if (lock == null) {
+                listener.onFailure(new OpenSearchException("another processor is holding a lock on the resource. Try again later"));
+                return;
+            }
+            try {
+                Datasource datasource = datasourceFacade.getDatasource(request.getName());
+                if (datasource == null) {
+                    throw new ResourceNotFoundException("no such datasource exist");
```

Comment:
Why doesnt this use listener.onfailure like line 77?

Reply:
single line vs double line. I though single line is simpler.
```
if (datasource == null){
  listener.onFailure(new ResourceNotFoundException("no such datasource exist");
  return;
}
```

Reply:
doExecute throwing an exception which propagate it back up its call trace shouldn't happen. If that happens, the user will never receive a response and timed out eventually.